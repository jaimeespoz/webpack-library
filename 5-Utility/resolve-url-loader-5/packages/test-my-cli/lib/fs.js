'use strict';

const {join} = require('path');
const {existsSync, readdirSync, stat, readFile, writeFile, symlink, unlink} = require('fs');
const {isAbsolute} = require('path');
const {promisify} = require('es6-promisify');
const mkdirp = require('mkdirp');
const rimraf = require('rimraf');
const recursiveReaddir = require('recursive-readdir');

const sequence = require('promise-compose');
const {conditional, constant} = require('./promise');

const noop = () => {};

const getStat = sequence(
  conditional((v) => (typeof v === 'string') && isAbsolute(v) && existsSync(v)),
  (v) => v && promisify(stat)(v)
);

exports.safeRemoveDirSync = (baseDir, subDir) =>
  subDir
    .split(/[\\\/]/)
    .map((_, i, arr) => join(baseDir, ...arr.slice(0, arr.length - i)))
    .forEach((directory, i) => {
      if ((i === 0) || !readdirSync(directory).length) {
        rimraf.sync(directory, {glob: false});
      }
    });

exports.testIsFile = sequence(
  getStat,
  (stats) => stats && stats.isFile()
);

exports.testIsDir = sequence(
  getStat,
  (stats) => stats && stats.isDirectory()
);

exports.testIsEmpty = sequence(
  conditional(exports.testIsDir),
  (directory) => directory && recursiveReaddir(directory),
  (files) => files && (files.length === 0)
);

const withCatch = (log, text) => (fn) => {
  log(text);
  return fn().catch((error) => {
    console.error(`Can't ${text}`);
    console.error(error.stack.split('\n')[1] || '-no stack trace-');
    throw error;
  });
};

exports.MkDirOp = class MkDirOp {
  constructor({path, log = noop}) {
    this.path = path;
    this.log = log;
  }

  exec() {
    return withCatch(this.log, `mkdir: ${this.path}`)(
      () => mkdirp(this.path)
        .then((created) => {this.created = created;})
        .then(constant(this))
    );
  }

  undo() {
    return withCatch(this.log, `rmdir: ${this.created || '-nothing-'}`)(
      () => conditional(exports.testIsEmpty)(this.created)
        .then((created) => created && promisify(rimraf)(created, {glob: false}))
        .then(constant(this))
    );
  }
};

exports.CleanOp = class CleanOp {
  constructor({path, log = noop}) {
    this.path = path;
    this.log = log;
  }

  exec() {
    return withCatch(this.log, `clean: ${this.path}`)(
      () => conditional(exports.testIsDir)(this.path)
        .then((dir) => dir && promisify(rimraf)([dir, '**', '*'].join('/')))
        .then(constant(this))
    );
  }

  undo() {
    return this;
  }
};

exports.SymLinkOp = class SymLinkOp {
  constructor({srcPath, destPath, log = noop}) {
    this.srcPath = srcPath;
    this.destPath = destPath;
    this.log = log;
  }

  exec() {
    return withCatch(this.log, `symlink: ${this.destPath} => ${this.srcPath}`)(
      () => promisify(symlink)(this.srcPath, this.destPath, 'junction')
        .then(constant(this))
    );
  }

  undo() {
    return withCatch(this.log, `unlink: ${this.destPath}`)(
      () => promisify(unlink)(this.destPath)
        .then(constant(this))
    );
  }
};

exports.CopyOp = class CopyOp {
  constructor({srcPath, destPath, log = noop}) {
    this.srcPath = srcPath;
    this.destPath = destPath;
    this.log = log;
  }

  exec() {
    return withCatch(this.log, `copy: ${this.srcPath} => ${this.destPath}`)(
      () => promisify(readFile)(this.srcPath)
        .then((content) => promisify(writeFile)(this.destPath, content))
        .then(constant(this))
    );
  }

  undo() {
    return withCatch(this.log, `rm: ${this.destPath}`)(
      () => promisify(unlink)(this.destPath)
        .then(constant(this))
    );
  }
};

exports.WriteOp = class WriteOp {
  constructor({content, destPath, log = noop}) {
    this.content = content;
    this.destPath = destPath;
    this.log = log;
  }

  exec() {
    return withCatch(this.log, `write: [text] => ${this.destPath}`)(
      () => promisify(writeFile)(this.destPath, this.content)
        .then(constant(this))
    );
  }

  undo() {
    return withCatch(this.log, `rm: ${this.destPath}`)(
      () => promisify(unlink)(this.destPath)
        .then(constant(this))
    );
  }
};
