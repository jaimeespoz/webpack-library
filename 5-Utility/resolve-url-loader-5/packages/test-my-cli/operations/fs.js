'use strict';

const {basename, join, dirname} = require('path');
const compose = require('compose-function');
const {keys, values, entries} = Object;

const joi = require('../lib/joi');
const {assertInLayer} = require('../lib/assert');
const {lens, sequence, mapSerial, mapParallel} = require('../lib/promise');
const {operation, assertInOperation} = require('../lib/operation');
const {testIsFile, testIsDir, MkDirOp, CleanOp, SymLinkOp, CopyOp, WriteOp} = require('../lib/fs');

const NAME = basename(__filename).slice(0, -3);

const mergeUndos = (layer) => (undos) => {
  const {register} = layer;
  undos.forEach(register);
  return layer;
};

const hashToTuple = (hash) => (layer) =>
  entries(hash).map(([k, v]) => {
    const {root} = layer;
    return [
      layer,
      (typeof v === 'function') ? v(layer) : v,
      join(root, k)
    ];
  });

const tupleToOp = ([{root}, srcPath, destPath], _, log) => {
  switch (true) {
    case (srcPath === null):
      return (destPath === root) ?
        [] :
        [new MkDirOp({path: destPath, log}), new CleanOp({path: destPath, log})];

    case (typeof srcPath === 'string'):
      if (destPath.startsWith(root)) {
        return Promise
          .all([testIsFile(srcPath).catch(() => console.error(srcPath)), testIsDir(srcPath)])
          .then(([isSrcFile, isSrcDir]) =>
            isSrcDir ? new SymLinkOp({srcPath, destPath, log}) :
              isSrcFile ? new CopyOp({srcPath, destPath, log}) :
                new WriteOp({content: srcPath, destPath, log})
          )
          .then((op) => {
            const destDir = dirname(destPath);
            return (destDir === root) ?
              op :
              [new MkDirOp({path: destDir, log}), op];
          });
      }
      throw new Error(`Given path is outside the root: "${destPath}"`);

    default:
      throw new Error(`Expected key to be null|string, saw ${typeof srcPath}`);
  }
};

const flatten = (array) =>
  array.reduce((r, element) => r.concat(element), []);

const reverse = (array) =>
  [...array].reverse();

exports.schema = {
  debug: joi.debug().optional()
};

/**
 * Given a hash of keys will create directories and files that can be rolled back when the layer
 * completes.
 *
 * All keys are a path to a file or directory.
 *
 * - A value of `null` implies the key is a path that should be created and cleaned.
 * - A value that is an existing file implies the key should be a copy of that file.
 * - A value that is an existing directory implies the key should be a symlink to that directory.
 * - Any other `string` is direct file content and to be found in the file which is the key.
 *
 * @param {object} hash A hash of file system items
 * @return {function(object):Promise} A pure async function of the test context
 */
exports.create = (hash) => {
  joi.assert(
    keys(hash),
    joi.array().items(
      joi.path().relative().required()
    ).required(),
    'hash where keys are file or directory paths'
  );
  joi.assert(
    values(hash),
    joi.array().items(
      joi.alternatives().try(
        joi.any().only(null),
        joi.func(),
        joi.string()
      ).required()
    ).required(),
    'hash where values are null|existing-file-path|existing-directory-path|file-content'
  );

  return compose(operation(NAME), lens('layer', 'layer'), sequence)(
    assertInOperation(`misuse: ${NAME}() somehow escaped the operation`),
    assertInLayer(`${NAME}() must be used within layer()`),
    compose(lens('*', mergeUndos), sequence)(
      hashToTuple(hash),
      lens('*', null)((list, _, log) => log(`${list.length} tuples`)),
      mapParallel(tupleToOp),
      flatten,
      lens('*', null)((list, _, log) => log(`${list.length} operations`)),
      mapSerial((op) => op.exec()),
      reverse,
      mapParallel((op) => () => op.undo())
    )
  );
};
