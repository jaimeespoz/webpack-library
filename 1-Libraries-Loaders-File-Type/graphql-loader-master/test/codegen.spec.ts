import { runFixture, TestRunError } from "./runner";
import { readFileSync } from "fs";

describe("codegen", function() {
  it("codegen", async function() {
    const document = await runFixture("codegen");
    expect(document).toMatchSnapshot();

    const declarationFile = await readFileSync(
      "./test/fixtures/codegen/query.graphql.d.ts",
      null,
    );
    expect(declarationFile.toString()).toMatchSnapshot();
  });

  it("codegen with default export", async function() {
    const document = await runFixture("codegen-export-default");
    expect(document).toMatchSnapshot();

    const declarationFile = await readFileSync(
      "./test/fixtures/codegen-export-default/query.graphql.d.ts",
      null,
    );
    expect(declarationFile.toString()).toMatchSnapshot();
  });
});
