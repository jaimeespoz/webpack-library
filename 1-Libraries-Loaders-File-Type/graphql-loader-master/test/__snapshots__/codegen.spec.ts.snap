// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen codegen 1`] = `
Object {
  "__esModule": true,
  "default": Object {
    "query": "query Films($k: ID!) {
  film(id: $k) {
    title
  }
}
",
  },
}
`;

exports[`codegen codegen 2`] = `
"/* tslint:disable */
//  This file was automatically generated and should not be edited.

export interface FilmsQueryVariables {
  k: string,
}

export interface FilmsQuery {
  film:  {
    // The title of this film.
    title: string | null,
  } | null,
}

export interface Spec<Result extends object, Input extends object> {
  query: string
  result?: Result
  variables?: Input
}

declare const Dummy: Spec<FilmsQuery, FilmsQueryVariables>
export default Dummy
"
`;

exports[`codegen codegen with default export 1`] = `
Object {
  "__esModule": true,
  "default": Object {
    "__esModule": true,
    "default": Object {
      "query": "query Films($k: ID!) {
  film(id: $k) {
    title
  }
}
",
    },
  },
}
`;

exports[`codegen codegen with default export 2`] = `
"/* tslint:disable */
//  This file was automatically generated and should not be edited.

export interface FilmsQueryVariables {
  k: string,
}

export interface FilmsQuery {
  film:  {
    // The title of this film.
    title: string | null,
  } | null,
}

export interface Spec<Result extends object, Input extends object> {
  query: string
  result?: Result
  variables?: Input
}

declare const Dummy: Spec<FilmsQuery, FilmsQueryVariables>
export default Dummy
"
`;
